src/controllers/controllerengine.cpp

```diff
@@ -98,30 +98,44 @@ void ControllerEngine::callFunctionOnObjects(QList<QString> scriptFunctionPrefix
     }
 }
 
-/* -------- ------------------------------------------------------
-Purpose: Resolves a function name to a QScriptValue including
-            OBJECT.Function calls
-Input:   -
-Output:  -
--------- ------------------------------------------------------ */
-QScriptValue ControllerEngine::resolveFunction(const QString& function) const {
+/* ------------------------------------------------------------------
+Purpose: Turn a snippet of JS into a QScriptValue function.
+         Wrapping it in an anonymous function allows any JS that
+         evaluates to a function to be used in MIDI mapping XML files
+         and ensures the function is executed with the correct
+         'this' object.
+Input:   QString snippet of JS that evaluates to a function
+Output:  QScriptValue of JS snippet wrapped in an anonymous function
+------------------------------------------------------------------- */
+QScriptValue ControllerEngine::wrapFunctionCode(const QString& codeSnippet) {
+    QScriptValue wrappedFunction;
+
     QHash<QString, QScriptValue>::const_iterator i =
-            m_scriptValueCache.find(function);
-    if (i != m_scriptValueCache.end()) {
-        return i.value();
-    }
+            m_scriptWrappedFunctionCache.find(codeSnippet);
 
-    QScriptValue object = m_pEngine->globalObject();
-    QStringList parts = function.split(".");
+    if (i != m_scriptWrappedFunctionCache.end()) {
+        wrappedFunction = i.value();
+    } else {
+        QScriptValue function = m_pEngine->evaluate(codeSnippet);
+        if (!syntaxIsValid(codeSnippet)) {
+            return m_pEngine->evaluate("(function () {})");
+        }
+        if (checkException()) {
+            return m_pEngine->evaluate("(function () {})");
+        }
 
-    for (int i = 0; i < parts.size(); i++) {
-        object = object.property(parts.at(i));
-        if (!object.isValid()) {
-            break;
+        QString wrappedCode = "(function (channel, control, value, status, group) { (" +
+                              codeSnippet + ")(channel, control, value, status, group); })";
+        wrappedFunction = m_pEngine->evaluate(wrappedCode);
+
+        if (checkException()) {
+            // There will be an exception if the codeSnippet didn't evaluate to a function
+            return m_pEngine->evaluate("(function () {})");
         }
+
+        m_scriptWrappedFunctionCache[codeSnippet] = wrappedFunction;
```

Comment:
I thought about that, but I realized it could be possible for a failed lookup to work the next time if it references script variables that the script has changed between calls.